compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs

```diff
@@ -136,21 +164,86 @@ fn insert_required_predicates_to_be_wf<'tcx>(
+                        // impl<'a, R: 'a> Trait<'a> for RGen<R> {
+                        //     type Assoc = R; // works with () too
+                        // }
+                        // ```
+                        //
+                        // Visiting Node, we walk the fields and find a Var. Var has an explicit
+                        //     R                         : 'var.
+                        // Node finds this in check_explicit_predicates.
+                        // Node substitutes its type parameters for Var through, and gets an inferred
+                        //     <T as Trait<'node>>::Assoc: 'node.
+                        // Visiting Var, we walk the fields and find Node, for which there us an
+                        // unsubstituted predicate in the global map. So Var gets
+                        //     <RGen<R> as Trait<'var>>::Assoc: 'var
+                        // But Node contains a Var. So Node gets
+                        //     <RGen<<T as Trait<'node>>::Assoc> as Trait<'node>>::Assoc 'node
+                        // Var gets
+                        //     <RGen<<RGen<R> as Trait<'var>>::Assoc> as Trait<'var>>::Assoc: 'var
+                        //
+                        // Etc. This goes on forever. The fact that RGen<R>::Assoc *is* R is
+                        // irrelevant. It is simply that both types find a way to add a bigger
+                        // predicate each time.
+                        //
+                        // The outlives requirements propagate through the crate's types like a
+                        // graph walk, so to detect this cycle we can just track visited nodes in
+                        // our branch by pushing them on a stack when we move through the graph.
+                        // An edge move is when a type picks up a predicate from one of its fields
+                        // that it hasn't seen before. We store the stacks alongside the predicates,
+                        // so they carry their provenance with them through the current and
+                        // subsequent rounds of crate-wide inference.
+                        //
+                        // Try: RUSTC_LOG=rustc_hir_analysis::outlives=debug \
+                        //      rustc +stage1 src/test/ui/rfc-2093-infer-outlives/issue-102966.rs 2>&1 \
+                        //      | rg '(grew|cycle)'
+                        //
+                        // We do not treat a type with an explicit bound as the first in the visit
+                        // stack. So in the example, Node appears first in the stack, and each of
+                        // Node and Var will get a new bound constraining an ::Assoc projection
+                        // before the cycle is cut at Node. If Var were already in the visit stack,
+                        // it would not receive the projection bound, which is something it needs.
```

Comment:
This is the main thing I would like some eyes on. I think it's right for Var to get a substituted-through version of its own explicit bound here, because Var doesn't know what `RGen<R>::Assoc` is, but whatever it is, Node does need it to be `: 'node`. It seems the cautious approach to add the bound if we're not going to find out what the projection resolves to.

But I don't know and don't see much guidance in the RFC or its tests, as this situation wasn't really contemplated. Can anyone think of a test that might illuminate the correctness of this?

Tangentially, this PR has more tracking of the source of inferred outlives than before. Are there error messages we can improve with this information? Like the example of a "long range error" message [in the original RFC](https://rust-lang.github.io/rfcs/2093-infer-outlives.html#long-range-errors-and-why-they-are-considered-unlikely).